src/Catalyst.Core.Modules.Dfs/DfsService.cs

```diff
@@ -0,0 +1,663 @@
+
+        /// <summary>
+        ///     A synchronous stop.
+        /// </summary>
+        /// <remarks>
+        ///     Calls <see cref="StopAsync" /> and waits for it to complete.
+        /// </remarks>
+        public void Stop()
+        {
+            Log.Debug("stopping");
+            try
+            {
+                var tasks = _stopTasks.ToArray();
+                _stopTasks = new ConcurrentBag<Func<Task>>();
+                foreach (var task in tasks)
+                {
+                    task().ConfigureAwait(false).GetAwaiter().GetResult();
+                }
+            }
+            catch (Exception e)
+            {
+                Log.Error("Failure when stopping the engine", e);
+            }
+        }
+
+#pragma warning disable VSTHRD100 // Avoid async void methods
+        /// <summary>
+        ///     Fired when a peer is discovered.
+        /// </summary>
+        /// <param name="sender"></param>
+        /// <param name="peer"></param>
+        /// <remarks>
+        ///     Registers the peer with the <see cref="SwarmService" />.
+        /// </remarks>
+        private void OnPeerDiscovered(object sender, Peer peer)
+#pragma warning restore VSTHRD100 // Avoid async void methods
+        {
+            try
+            {
+                var swarm = SwarmService;
+                swarm.RegisterPeer(peer);
+            }
+            catch (Exception ex)
+            {
+                Log.Warning("failed to register peer " + peer, ex);
+
+                // eat it, nothing we can do.
+            }
+        }
+
+        /// <summary>
+        ///     @TODO this should reeally go lower down in the kernal when we load keystores
+        ///     Provides access to the <see cref="KeyStoreService" />.
+        /// </summary>
+        /// <param name="cancel">
+        ///     Is used to stop the task.  When cancelled, the <see cref="TaskCanceledException" /> is raised.
+        /// </param>
+        /// <returns>
+        ///     A task that represents the asynchronous operation. The task's result is
+        ///     the <see cref="KeyStoreService" />.
+        /// </returns>
+        public async Task<IKeyStoreService> KeyChainAsync(CancellationToken cancel = default)
+        {
+            // TODO: this should be a LazyAsync property.
+            if (_keyStoreService != null)
+            {
+                return _keyStoreService;
+            }
+            
+            lock (this)
+            {
+                if (_keyStoreService == null)
+                {
+                    _keyStoreService = new KeyStoreService(Options);
+                }
+            }
+
+            await _keyStoreService.SetPassphraseAsync(_passphrase, cancel).ConfigureAwait(false);
+
+            // Maybe create "self" key, this is the local peer's id.
+            var _ = await _keyStoreService.FindKeyByNameAsync("self", cancel).ConfigureAwait(false) ??
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Remove the unused local variable '_'.](https://app.codacy.com/manual/catalyst-network/Catalyst.Framework/pullRequest?prid=4832783)